### ISSUE TITLE
Long build queue, builds not starting

### ISSUE THREAD

**Issue by mbrodala** [Shippable/support#4917]
Right now we experience builds taking a lot of time in the build queue before they are started. Our build nodes are free so builds should be able to start instantly like usual.

You can see this right now with build 200 of project `58dd331330240a070076c85e`. Also build 478, 478 and 480 of project `59c3af5636154a070093df38`.

Can you have a look at this?

**Comment by mbrodala** [comment]
Some of the builds where started in the meantime but the queue duration shown on the build page is incorrect. E.g. build 478 of project `59c3af5636154a070093df38` did wait 2 minutes in the queue (which is extremely long already) but in fact it where roughly  20 minutes instead.

**Comment by rageshkrishna** [comment]
Hi @mbrodala. The queuing issue was due to a strange database connectivity issue that affected some of our internal APIs. Some API calls from our microservices were timing out due to this, resulting in multiple retries before a job could get processed. This should be resolved now as we are no longer seeing timeouts on these calls.

I'm not yet sure what could have caused an incorrect queue duration on these builds. I'll check and update you once I have a root cause.

**Comment by mbrodala** [comment]
Yes, it seems like builds start instantly again now, thanks for the fix.

I'm looking forward to more details on this issue

**Comment by rageshkrishna** [comment]
A few of our database connections ended up in a "stuck" state (for lack of a better explanation) on postgres. We tried to cancel the queries in the db but it didn't help to unblock the connections.

These db connections in the internal APIs were unable to perform any more work after this point. As a result, API calls began to slow down and queue up due to a lack of active connections to the database. We had to kill the APIs so they could establish new connections. This instantly resolved the issue.

We don't have a root cause yet but if I were to make a bet it would be on a temporary glitch in the network.

**Comment by mbrodala** [comment]
Thanks for the explanation, I got that. Do you have an idea how to prevent such issues in the future or at least add helpful countermeasures in case it happens again?

**Comment by rageshkrishna** [comment]
Yes, this had already tripped our latency alarms well before anyone noticed so we were already investigating. However, it was a novel situation and we spent more time than we should have trying to understand exactly what was going on. It was further complicated by the fact that some routine database maintenance activities were also triggered around the same time and we were trying to determine whether that had any part in causing the problem.

Long story short, now that we have experienced this situation we do know how to resolve it quickly if it happens again.

**Comment by mbrodala** [comment]
Alright, thanks for the insights! :-)
